src/psfmachine/machine.py

```diff
@@ -837,289 +635,197 @@ def build_time_model(self, plot=False, downsample=False, split_time_model=False)
+            self.source_mask.multiply(self.dra),
+            self.source_mask.multiply(self.ddec),
+            self.source_mask.multiply(self.uncontaminated_source_mask.todense()),
         )
         dx = dx.data * u.deg.to(u.arcsecond)
         dy = dy.data * u.deg.to(u.arcsecond)
+        # uncontaminated pixels mask
+        uncontaminated_pixels = uncontaminated_pixels.data
 
-        A_c = _make_A_cartesian(
-            dx,
-            dy,
-            n_knots=self.n_time_knots,
-            radius=self.time_radius,
-            spacing=self.cartesian_knot_spacing,
+        # add other vectors if asked, centroids or poscorrs
+        splits = get_breaks(self.time)
+        self.poscorr_filter_size = (
+            0.1 if self.poscorr_filter_size < 0.5 else self.poscorr_filter_size
         )
-        A2 = sparse.vstack([A_c] * time_binned.shape[0], format="csr")
-        # Cartesian spline with time dependence
-        if hasattr(self, "pos_corr1") and self.time_corrector in [
-            "pos_corr",
-            "centroid",
-        ]:
-            # Cartesian spline with poscor dependence
-            A3 = _combine_A(A2, poscorr=[poscorr1_binned, poscorr2_binned])
+        if self.time_corrector == "polynomial":
+            other_vectors = None
         else:
-            # Cartesian spline with time dependence
-            A3 = _combine_A(A2, time=time_binned)
-
-        prior_sigma = np.ones(A3.shape[1]) * 10
-        prior_mu = np.zeros(A3.shape[1])
-
-        # fit the time model for each segment
-        # use user input
-        if isinstance(split_time_model, (np.ndarray, list)):
-            # we make sure first and last index are included and sorted
-            splits = np.unique(np.append(split_time_model, [0, len(self.time)]))
-        elif isinstance(split_time_model, bool):
-            # we find the splits in data
-            if split_time_model:
-                splits = get_breaks(self.time)
-            else:
-                splits = np.array([0, len(self.time)])
-
-        seg_time_model_w, pix_mask_k = [], []
-        # we iterate over segements
-        for bk in range(len(splits) - 1):
-            # find the right mask that select the binned times andd flux to fit the
-            # time model
-            seg_mask = (time_binned[:, 0] >= time_original[splits[bk]]) & (
-                time_binned[:, 0] < time_original[splits[bk + 1] - 1]
+            if self.time_corrector == "pos_corr" and hasattr(self, "pos_corr1"):
+                mpc1 = np.nanmedian(self.pos_corr1, axis=0)
+                mpc2 = np.nanmedian(self.pos_corr2, axis=0)
+                # other_vectors =
+            elif self.time_corrector == "centroid" and hasattr(self, "ra_centroid"):
+                mpc1 = self.ra_centroid.to("arcsec").value / 4
+                mpc2 = self.dec_centroid.to("arcsec").value / 4
+
+            # smooth the vectors
+            mpc1_smooth, mpc2_smooth = smooth_vector(
```

Comment:
We shouldn't use a Gaussian smooth here we know it's noisy, we should replace it with a spline fit. I think we can have that be part of the 'cleaning' in perturbation matrix

Reply:
I think the `other_vectors` should be already smooth before being passed to `perturbation` because we could pass other types of vectors that don't need "smoothing" or "cleaning". I think the "cleaning" inside `perturbation` only attends to removing the degeneracy in the basis we are using, aka removing the time trend.

Reply:
I'll add an auxiliary function in `utils` to apply a spline smoothing to a vector vector `y`.

Reply:
Here's an example using the new `spline_smooth()` function in `utils.py`:

![image](https://user-images.githubusercontent.com/10449555/162816730-9114c340-e35c-474a-ba91-8c7b48f26b89.png)

Reply:
I improved the `spline_smooth` function. it should work with any type of vector now defining the number of knots and breaks if wanted.